src/drivers/ammo-driver.js

```diff
@@ -1,10 +1,174 @@
+      Ammo = result;
+      this.epsilon = worldConfig.epsilon || EPS;
+      this.debugDrawMode = worldConfig.debugDrawMode || THREE.AmmoDebugConstants.NoDebug;
+      this.maxSubSteps = worldConfig.maxSubSteps || 4;
+      this.fixedTimeStep = worldConfig.fixedTimeStep || 1 / 60;
+      this.collisionConfiguration = new Ammo.btDefaultCollisionConfiguration();
+      this.dispatcher = new Ammo.btCollisionDispatcher(this.collisionConfiguration);
+      this.broadphase = new Ammo.btDbvtBroadphase();
+      this.solver = new Ammo.btSequentialImpulseConstraintSolver();
+      this.physicsWorld = new Ammo.btDiscreteDynamicsWorld(
+        this.dispatcher,
+        this.broadphase,
+        this.solver,
+        this.collisionConfiguration
+      );
+      this.physicsWorld.setGravity(
+        new Ammo.btVector3(0, worldConfig.hasOwnProperty("gravity") ? worldConfig.gravity : -9.8, 0)
+      );
+      this.physicsWorld.getSolverInfo().set_m_numIterations(worldConfig.solverIterations);
+      resolve();
+    });
+  });
+};
+
+/* @param {Ammo.btCollisionObject} body */
+AmmoDriver.prototype.addBody = function(body, group, mask) {
+  this.physicsWorld.addRigidBody(body, group, mask);
+  this.els[Ammo.getPointer(body)] = body.el;
+};
+
+/* @param {Ammo.btCollisionObject} body */
+AmmoDriver.prototype.removeBody = function(body) {
+  this.physicsWorld.removeRigidBody(body);
+  this.removeEventListener(body);
+  delete this.els[Ammo.getPointer(body)];
+};
+
+AmmoDriver.prototype.updateBody = function(body) {
+  if (this.els.hasOwnProperty(Ammo.getPointer(body))) {
+    this.physicsWorld.updateSingleAabb(body);
+  }
+};
+
+/* @param {number} deltaTime */
+AmmoDriver.prototype.step = function(deltaTime) {
+  this.physicsWorld.stepSimulation(deltaTime, this.maxSubSteps, this.fixedTimeStep);
+
+  const numManifolds = this.dispatcher.getNumManifolds();
+  for (let i = 0; i < numManifolds; i++) {
+    const persistentManifold = this.dispatcher.getManifoldByIndexInternal(i);
+    const numContacts = persistentManifold.getNumContacts();
+    const body0ptr = Ammo.getPointer(persistentManifold.getBody0());
+    const body1ptr = Ammo.getPointer(persistentManifold.getBody1());
+    const key = body0ptr + "_" + body1ptr;
+    let collided = false;
+    for (let j = 0; j < numContacts; j++) {
+      const manifoldPoint = persistentManifold.getContactPoint(j);
+      const distance = manifoldPoint.getDistance();
+      if (body0ptr !== body1ptr && distance <= this.epsilon) {
+        collided = true;
+        break;
+      }
+    }
+    if (collided) {
+      if (!this.collisions.hasOwnProperty(key)) {
+        this.collisions[key] = [body0ptr, body1ptr];
+        if (this.eventListeners.indexOf(body0ptr) !== -1) {
+          this.els[body0ptr].emit("collide", { targetEl: this.els[body1ptr] });
+        }
+        if (this.eventListeners.indexOf(body1ptr) !== -1) {
+          this.els[body1ptr].emit("collide", { targetEl: this.els[body0ptr] });
```

Comment:
It seems unfortunate to ask Ammo to compute all of the precise contact points between two bodies and then throw it all away except the bit of information "did they collide at all." Is it possible to either use Ammo in a way that does less work?

Reply:
The following is gleamed from various sources including but not limited to:
https://pybullet.org/Bullet/phpBB3/viewtopic.php?f=9&t=3997
https://pybullet.org/Bullet/phpBB3/viewtopic.php?f=9&t=12116
https://pybullet.org/Bullet/phpBB3/viewtopic.php?f=9&t=11968
https://github.com/AndresTraks/BulletSharp/wiki/Collision-Callbacks-and-Triggers
(Unfortunately the actual Bullet Physics wiki has been down for some time and doesn't look like its coming back anytime soon, so the actual documentation: http://bulletphysics.org/mediawiki-1.5.8/index.php/Collision_Callbacks_and_Triggers) is unavailable. 

There are 5 ways to "check" collisions in Bullet:
1. Iterate over contacts. This is what is currently being done here.
2. Use btGhostObjects. This seems to be the only reasonable alternative for our needs, though I am very unclear on how to use it. I believe for btRigidBody that you care about collisions for, create a btGhostObject that will keep track of other objects that are overlapping with this object. (By default this only checks AABB, but I think you can have it use the collision shape of the parent object? It may require overriding the btGhostObject class. Also, I'm not sure if this is limited only to broadphase collision or not).
3. ContactTest - On demand, get a list of all contacts. This is duplicating work already done every frame and therefore only meant to be used when you only occasionally need to query the collisions in a scene.
4. ContactPairTest - Allows querying of specific pair of objects.
5. ContactCallbacks - My understanding is these are not meant to be used for checking collision, as much as they are meant to be used to effectively replace existing collision behavior with your own implementation.  

My main reason for iterating over the contacts was that it was the most straight forward way to achieve the needs of the system.

Reply:
Oh, yeah, the way that we're checking collisions doesn't concern me so much. What I'm really hoping is that we could tell Bullet to do the physics simulation in some way such that less information comes out the other end, but in return it's faster. I don't have any great reason to believe it's possible, but that would be my dream.